Python/Product/Ipc.Json/ProtocolReader.cs

```diff
@@ -0,0 +1,121 @@
+        /// dispose of the stream.
+        /// </summary>
+        public async Task<string> ReadHeaderLineAsync() {
+            // Keep reading into the buffer until it contains the '\r\n'.
+            int searchStartPos = 0;
+            int newLineIndex;
+            while ((newLineIndex = IndexOfNewLineInBuffer(searchStartPos)) < 0) {
+                searchStartPos = Math.Max(0, _buffer.Count - 1);
+
+                // Keep reading until we get data
+                int readCount;
+                while ((readCount = await ReadIntoBuffer()) == 0) {
+                }
+            }
+
+            var line = _buffer.Take(newLineIndex).ToArray();
+            _buffer.RemoveRange(0, newLineIndex + 2);
+            var text = Encoding.ASCII.GetString(line);
+            return text;
+        }
+
+        /// <summary>
+        /// Reads all bytes from the current position to the end of the
+        /// stream asynchronously.
+        /// </summary>
+        public async Task<byte[]> ReadToEndAsync() {
+            int read;
+            while ((read = await ReadIntoBuffer()) > 0) {
+            }
+
+            var all = _buffer.ToArray();
+            _buffer.Clear();
+            return all;
+        }
+
+        /// <summary>
+        /// Reads a number of bytes asynchronously from the current stream.
+        /// </summary>
+        /// <param name="byteCount">Number of bytes to read.</param>
+        /// <remarks>
+        /// May return fewer bytes than requested.
+        /// </remarks>
+        public async Task<byte[]> ReadContentAsync(int byteCount) {
+            if (byteCount < 0) {
+                throw new ArgumentOutOfRangeException(nameof(byteCount), byteCount, "Value cannot be negative.");
+            }
+
+            while (_buffer.Count < byteCount) {
+                var addedByteCount = await ReadIntoBuffer();
+                if (addedByteCount == 0) {
+                    break;
+                }
+            }
+
+            var actualCount = Math.Min(byteCount, _buffer.Count);
+            var result = _buffer.Take(actualCount).ToArray();
+            _buffer.RemoveRange(0, actualCount);
+            return result;
+        }
+
+        private int IndexOfNewLineInBuffer(int searchStartPos) {
+            for (int i = searchStartPos; i < _buffer.Count - 1; i++) {
+                if (_buffer[i] == 13 && _buffer[i + 1] == 10) {
+                    return i;
+                }
+            }
+            return -1;
+        }
+
+        /// <summary>
+        /// Reads bytes from the stream into the buffer, in chunks.
+        /// </summary>
+        /// <returns>Number of bytes that were added to the buffer.</returns>
+        private async Task<int> ReadIntoBuffer() {
+            var temp = new byte[1024];
+            var read = await _stream.ReadAsync(temp, 0, temp.Length);
+            _buffer.AddRange(temp.Take(read).ToArray());
```

Comment:
Performance impact should be minimal.  I will do this when I have enough time to rerun tests.